drivers/usb/device/usb_dc_native_posix_adapt.c

```diff
@@ -0,0 +1,452 @@
+
+	return 0;
+}
+
+static void handle_usbip_submit(int connfd, struct usbip_header *hdr)
+{
+	struct usbip_submit *req = &hdr->u.submit;
+	int read;
+
+	SYS_LOG_DBG("");
+
+	read = recv(connfd, req, sizeof(*req), 0);
+	if (read != sizeof(*req)) {
+		SYS_LOG_ERR("recv() failed: %s", strerror(errno));
+		return;
+	}
+
+	usbip_header_dump((void *)hdr);
+
+	if (ntohl(hdr->common.ep) == 0) {
+		handle_usb_control(hdr);
+	} else {
+		handle_usb_data(hdr);
+	}
+}
+
+static void handle_usbip_unlink(int connfd, struct usbip_header *hdr)
+{
+	struct usbip_unlink *req = &hdr->u.unlink;
+	u64_t setup_padding;
+	int read;
+
+	SYS_LOG_DBG("");
+
+	read = recv(connfd, req, sizeof(hdr->u), 0);
+	if (read != sizeof(hdr->u)) {
+		SYS_LOG_ERR("recv() failed: %s", strerror(errno));
+		return;
+	}
+
+	/* Read also padding */
+	read = recv(connfd, &setup_padding, sizeof(setup_padding), 0);
+	if (read != sizeof(setup_padding)) {
+		SYS_LOG_ERR("recv() failed: %s", strerror(errno));
+		return;
+	}
+
+	usbip_header_dump((void *)hdr);
+
+	/* TODO: unlink */
+}
+
+static int handle_import(const u8_t *desc, int connfd)
+{
+	struct op_common header = {
+		.version = htons(USBIP_VERSION),
+		.code = htons(OP_REP_IMPORT),
+		.status = 0,
+	};
+	char busid[32];
+
+	SYS_LOG_DBG("attach device");
+
+	if (recv(connfd, busid, 32, 0) != sizeof(busid)) {
+		SYS_LOG_ERR("recv() failed: %s", strerror(errno));
+		return errno;
+	}
+
+	if (send(connfd, &header, sizeof(header), 0) != sizeof(header)) {
+		SYS_LOG_ERR("send() header failed: %s", strerror(errno));
+		return errno;
+	}
+
+	send_device(desc, connfd);
+
+	return 0;
+}
+
+extern struct usb_desc_header __usb_descriptor_start[];
+
+void usbip_start(void)
+{
+	struct sockaddr_in srv;
+	unsigned char attached;
+	int listenfd, connfd;
+	const u8_t *desc;
+	int reuse = 1;
+
+	SYS_LOG_DBG("Starting");
+
+	/*
+	 * Do not use usb_get_device_descriptor();
+	 * to prevent double string fixing
+	 */
+	desc = (const u8_t *)__usb_descriptor_start;
+	if (!desc) {
+		SYS_LOG_ERR("Descriptors are not set");
+		exit(EXIT_FAILURE);
+	}
+
+	listenfd = socket(PF_INET, SOCK_STREAM, 0);
+	if (listenfd < 0) {
+		SYS_LOG_ERR("socket() failed: %s", strerror(errno));
+		exit(EXIT_FAILURE);
+	}
+
+	if (setsockopt(listenfd, SOL_SOCKET, SO_REUSEADDR,
+		       (const char *)&reuse, sizeof(reuse)) < 0) {
+		SYS_LOG_WRN("setsockopt() failed: %s", strerror(errno));
+	}
+
+	memset(&srv, 0, sizeof(srv));
+	srv.sin_family = AF_INET;
+	srv.sin_addr.s_addr = htonl(INADDR_ANY);
+	srv.sin_port = htons(USBIP_PORT);
+
+	if (bind(listenfd, (struct sockaddr *)&srv, sizeof(srv)) < 0) {
+		SYS_LOG_ERR("bind() failed: %s", strerror(errno));
+		exit(EXIT_FAILURE);
+	}
+
+	if (listen(listenfd, SOMAXCONN) < 0) {
+		SYS_LOG_ERR("listen() failed: %s", strerror(errno));
+		exit(EXIT_FAILURE);
+	}
+
+	while (true) {
+		struct sockaddr_in client_addr;
+		socklen_t client_addr_len = sizeof(client_addr);
+
+		connfd = accept(listenfd, (struct sockaddr *)&client_addr,
```

Comment:
Is this a blocking call? If so, you will block the whole Zephyr until somebody connects to the socket.
General note:
If the behavior you expect is for this Zephyr thread to yield to other Zephyr threads while waiting for the socket, then you probably want to have a loop with `select()`* with a `k_sleep(something)`
´* Or make it nonblocking and check for `EAGAIN` or `EWOULDBLOCK`, and, if so, k_sleep() for a small while before checking again.
(You may want to take a look at the native_posix ethernet or bluetooth drivers)

In general you should not make blocking calls to the host OS unless you really want to block the whole Zephyr.

The same applies to `recv()` below

Reply:
Thanks for the tip, I will change the accept() call to non-blocking.

Reply:
@aescolar Should this work?

```
--- a/drivers/usb/device/usb_dc_native_posix_adapt.c
+++ b/drivers/usb/device/usb_dc_native_posix_adapt.c
@@ -310,7 +310,7 @@ void usbip_start(void)
                exit(EXIT_FAILURE);
        }
 
-       listenfd = socket(PF_INET, SOCK_STREAM, 0);
+       listenfd = socket(PF_INET, SOCK_STREAM | SOCK_NONBLOCK, 0);
        if (listenfd < 0) {
                SYS_LOG_ERR("socket() failed: %s", strerror(errno));
                exit(EXIT_FAILURE);
@@ -343,8 +343,15 @@ void usbip_start(void)
                connfd = accept(listenfd, (struct sockaddr *)&client_addr,
                                &client_addr_len);
                if (connfd < 0) {
+                       if (errno == EAGAIN || errno == EWOULDBLOCK) {
+                               /* Non-blocking accept */
+                               sleep(1);
+
+                               continue;
+                       }
+
                        SYS_LOG_ERR("accept() failed: %s", strerror(errno));
-                       exit(1);
+                       exit(EXIT_FAILURE);
                }
 
                connfd_global = connfd;
```

Reply:
I thought that sleep() and k_sleep() are the same, from `lib/posix/sleep.c`:
```
unsigned sleep(unsigned int seconds)
{
	k_sleep(K_SECONDS(seconds));
	return 0;
}
```
So does it make a difference?
PS: I will change exit() to posix_exit()

Reply:
There you are not using the `sleep( )` from Zephyr's POSIX adaptation layer, but the one from the host OS. But I think you should not call either of them.

Reply:
I started to use k_sleep() and non-blocking socket ops instead of blocking socket ops and it works, how do you propose to use this without sleep?

Reply:
I'm not sure I follow your question.
My point was you should call `k_sleep()` instead of `sleep()`.